src/building/dock.c

```diff
@@ -218,23 +226,106 @@ int building_dock_get_destination(int ship_id, int exclude_dock_id, map_point *t
+    if ((dock_id = get_free_destination(ship_id, exclude_dock_id, tile, handled_goods))) {
         return dock_id;
-    } else if ((dock_id = get_queue_destination(ship_id, exclude_dock_id, SHIP_DOCK_REQUEST_2_FIRST_QUEUE, tile))) {
+    } else if ((dock_id = get_queue_destination(ship_id, exclude_dock_id, SHIP_DOCK_REQUEST_2_FIRST_QUEUE, tile, handled_goods))) {
         return dock_id;
     } else {
-        return get_queue_destination(ship_id, exclude_dock_id, SHIP_DOCK_REQUEST_4_SECOND_QUEUE, tile);
+        return get_queue_destination(ship_id, exclude_dock_id, SHIP_DOCK_REQUEST_4_SECOND_QUEUE, tile, handled_goods);
     }
 }
 
+// returns a list of goods that have been "handled" (i.e. the dock allowed for it to be traded) for each road network a ship has visited
+void building_dock_get_handled_goods(handled_good *handled_goods, int ship_id)
+{
+    figure* ship = figure_get(ship_id);
+
+    // zero out the array
+    memset(handled_goods, 0, sizeof(handled_good) * MAX_DOCKS);
+
+    int next_available_idx = 0;
+    // loop through the docks
+    for (int i = 0; i < MAX_DOCKS; i++) {
+        // check and see if the ship has visited this dock
+        if (!figure_trader_ship_docked_once_at_dock_by_num(ship, i)) {
+            continue;
+        }
+
+        // get the actual dock
+        int dock_id = city_buildings_get_working_dock(i);
+        if (!dock_id) {
+            continue;
+        }
+        building *dock = building_get(dock_id);
+
+        // see if we already have a handled_good pair for this road network
+        int added_handled_good = 0;
+        for (int j = 0; j < MAX_DOCKS; j++) {
+            handled_good *handled_good = &handled_goods[j];
+            if (handled_good->road_network_id == dock->road_network_id) {
+                // we've already visited a dock on this road network. add all the resources that this dock handles
+                for (int r = RESOURCE_MIN; r < RESOURCE_MAX; r++) {
+                    if (building_distribution_is_good_accepted(r - 1, dock)) {
+                        handled_good->goods[r - 1] = 1;
+                    }
+                }
+                added_handled_good = 1;
+                break;
+            }
+        }
+
+        if (!added_handled_good) {
+            // no handled_good found for this road network. use the next available one and add all the resources that the dock accepts
+            handled_good *handled_good = &handled_goods[next_available_idx];
+            handled_good->road_network_id = dock->road_network_id;
+            for (int r = RESOURCE_MIN; r < RESOURCE_MAX; r++) {
+                if (building_distribution_is_good_accepted(r - 1, dock)) {
+                    handled_good->goods[r - 1] = 1;
+                }
+            }
+            added_handled_good = 1;
+            next_available_idx++;
+        }
+    }
+}
+
+int building_dock_goods_handled(handled_good *handled_goods, building *dock, figure *ship)
```

Comment:
Again, this can be made `static`.

I'd also replace the function name with a simple `all_dock_goods_already_handled`.